distributed/worker_state_machine.py

```diff
@@ -1666,6 +1653,32 @@ def _select_keys_for_gather(
 
         return to_gather, total_nbytes
 
+    def _task_exceeds_transfer_limits(self, ts: TaskState, total_nbytes: int) -> bool:
+        """Return True if gathering the task together with `total_nbytes` of other data in the same message, False otherwise."""
+        if not self.transfer_incoming_bytes and not total_nbytes:
+            # When there is no other traffic, the top-priority task is fetched
+            # regardless of its size to ensure progress
+            return False
+
+        transfer_incoming_bytes_remaining = (
+            self.transfer_incoming_bytes_limit - self.transfer_incoming_bytes
+        )
+
+        if not total_nbytes:
+            # Ignore the message target for the top-priority task of each worker
+            # to ensure progress
```

Comment:
> This case is not covered by the `self.transfer_incoming_bytes == 0 and total_nbytes == 0` short-circuit above because we might already have some incoming transfers that have not completed, and here we're contemplating starting a new message transfer?

Exactly. We might have a data transfer from another worker in progress; thus, `self.transfer_incoming_bytes` would be > 0. So we want to observe our total limit of `self.transfer_incoming_bytes_limit`, but we still want to ignore the message size as the top-priority task on this remote worker might be larger than that.

Reply:
I slightly refactored the logic and improved the comment to make it easier to understand.